---
title: "neosex_figures"
format: html
editor: visual
---

```{r}
#| include: false
library(tidyverse)
library(ggrepel)
library(rnaturalearth)
library(ggsignif)
library(patchwork)
library(bcp)
library(DescTools)
library(ape)
library(ggtree)
```

## Map and PCA (Figure 2)

```{r}
regions <- read_csv("../maps/output/id_sex_region.csv", show_col_types = FALSE) %>%
  select(ID, Region_Narrow) %>%
  mutate(Region_Narrow = str_replace(pattern = " ", replacement = "_", Region_Narrow))

df <- read_tsv("../maps/data/metadata.txt", show_col_types = FALSE) %>%
  rename(lat = Lat_Dec_Deg, long = Long_Dec_Deg) %>%
  left_join(regions, by = "ID") %>%
  select(ID, lat, long, Region_Narrow, Sex)

eurafrica <- ne_countries(continent = c('africa', 'europe'), scale = "medium", returnclass = "sf")

map <- ggplot(data = eurafrica) +
  geom_sf(col = "white", size = 0.1, fill = "white") +
  geom_point(data = df, aes(x = long, y = lat, col = Region_Narrow), alpha = 6/10, size = 6, show.legend = FALSE) +
  geom_label_repel(data = df, aes(x = long, y = lat, col = Region_Narrow, label = ID),  box.padding = 0.75, max.overlaps = 100, seed = 2029, show.legend = FALSE) +
  labs(x = "", y = "") +
  coord_sf(xlim = c(-10.5,10.5), ylim = c(29.5,43.5)) +
  scale_x_continuous(breaks = c(-10, 0, 10)) +
  scale_y_continuous(breaks = c(30, 35, 40)) +
  scale_color_manual(values = MetBrewer::met.brewer("Egypt", 3)) + 
  theme_light() +
  theme(panel.grid = element_blank(), panel.background = element_rect(fill = "lightblue"))

file <- "//wsl.localhost/Ubuntu/home/thomas/research/neoZ/5_pca/autosomal/autosomal_pca.csv"
df <- read.csv(file)

pca <- df %>%
  ggplot(aes(PC1, PC2, label = ID, col = Region_Narrow)) +
  geom_point(size = 5, alpha = 0.6, show.legend = FALSE) +
  scale_color_manual(values = MetBrewer::met.brewer("Egypt", 3)) +
  geom_label_repel(box.padding = 0.75, max.overlaps = 100, seed = 2029, show.legend = FALSE) +
  labs(col = "Region", x = "PC1 (50.5%)", y = "PC2 (14.0%)") + 
  theme_light() +
  theme(plot.title = element_text(hjust = 0.5), 
        axis.ticks = element_blank(), axis.text = element_blank(), panel.grid = element_blank(),
        axis.title.y = element_text(vjust = 1))


map + pca + plot_annotation(tag_levels = 'A')
```

```{r}
ggsave("pngs/map_pca.map_annotation.png", width = 14, height = 6, dpi = 600)
```

### Heterozygosity and phylogenetic tree (Figure 3)

```{r}
#| include: false
library(tidyverse)
library(ggtree)
library(patchwork)
```

```{r}
file <- "//wsl.localhost/Ubuntu/home/thomas/research/neoZ/z_metadata/id_sex_region.csv"
meta <- read.csv(file) %>% rename(id = ID)

file <- "//wsl.localhost/Ubuntu/home/thomas/research/neoZ/012_4_fold_intergenic/4D_neosex_heterozygosity.csv"
df <- read.csv(file)

het <- df %>%
  left_join(meta, by = "id") %>%
  mutate(order = case_when(id == "PT_MI_7" ~ 1, id == "PT_MI_8" ~ 2, id == "PT_MI_61" ~ 4, id == "ES_MI_1647" ~ 3,
                            id == "PT_MI_86" ~ 5, id == "ES_MI_1680" ~ 6, id == "ES_MI_1682" ~ 7, id == "ES_MI_1683" ~ 8, 
                            id == "ES_MI_1684" ~ 9, id == "ES_MI_1686" ~ 10, id == "DZ_MI_1681" ~ 11, id == "DZ_MI_1685" ~ 13,
                            id == "DZ_MI_1624" ~ 12, id == "TN_MI_1619" ~ 14, id == "MA_MI_1620" ~ 15), 
         id = fct_reorder(id, order)) %>%
  filter(! str_detect(id, "MA")) %>%
  ggplot(aes(id, het, fill = Sex)) + 
  geom_col(position = "dodge") +
  labs(y = expression(H["4D"]), x = "", title = "Chromosome 14") +
  lims(y = c(0, 0.03)) +
  scale_fill_manual(values = c("#1f77b4", "#ff7f0e")) +
  facet_grid(~Region, scales = "free_x", space = "free_x")  +
  theme_light() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1), plot.title = element_text(hjust = 0.5))


file <- "//wsl.localhost/Ubuntu/home/thomas/research/neoZ/1_clean_vcfs/autosomal/trees/autosomes.contree"
aut <- read.tree(file)

file <- "//wsl.localhost/Ubuntu/home/thomas/research/neoZ/1_clean_vcfs/neoZ/trees/outgrouped_neo_sex.contree"
neosex <- read.tree(file)


aut <- ape::drop.tip(aut, "MA_MI_1620")
neosex <- ape::drop.tip(neosex, "MA_MI_1620")

p1 <- ggtree(aut) %<+% meta +
  geom_tippoint(aes(shape=Region, col = Sex), size = 3) +
  #geom_nodelab(size = 2, hjust = 1.2, vjust = -0.4) +
  scale_color_manual(values = c("#1f77b4", "#ff7f0e")) +
  ggtitle("Autosomes") +
  theme(plot.title = element_text(hjust = 0.5))

p2 <- ggtree(neosex) %<+% meta +
  geom_tippoint(aes(shape=Region, col = Sex), size = 3) +
  #geom_nodelab(size = 2, hjust = 1.2, vjust = -0.4) +
  scale_color_manual(values = c("#1f77b4", "#ff7f0e")) +
  ggtitle("Chromosome 14") +
  theme(plot.title = element_text(hjust = 0.5))

patchwork <- het / (p1 + p2 + plot_layout(guides = "collect"))
detection <- patchwork + plot_annotation(tag_levels = list(c('A', 'B')))
detection
```

## Change point detection (Figure 4)

```{r}
#| include: false
library(bcp)
```

```{r}
file <- "//wsl.localhost/Ubuntu/home/thomas/research/neoZ/z_metadata/id_sex_region.csv"
meta <- read.csv(file) %>% rename(id = ID) %>%
  select(id, class = Class)

file <- "//wsl.localhost/Ubuntu/home/thomas/research/neoZ/0_read_depth/neosex.coverage.50kb.csv"
df <- read_csv(file, show_col_types = FALSE) %>%
  left_join(meta, by = "id") %>%
  filter(class == "Female")

median_coverage <- df %>%
  group_by(mid) %>%
  summarise(median = median(median))
```

### Find the cutoff threshold 

```{r}
threshold_df <- tibble(threshold = numeric(), filtered = numeric())

n <- 1000

for (i in seq(n)) {
  
  threshold <- i / n
  filtered <- median_coverage %>% filter(median <= threshold) %>% nrow() / nrow(d)
  threshold_df <- threshold_df %>% rbind(tibble(threshold = threshold, filtered = filtered))
  
}

ggplot(threshold_df, aes(threshold, filtered)) +
  geom_line() +
  labs(y = "Proportion filtered out", x = "Coverage threshold") + 
  geom_hline(yintercept = 0.2, col = "red", linetype = "dashed") +
  scale_x_continuous(breaks = scales::breaks_pretty(n = 10)) +
  theme_light()
```

```{r}
filter(threshold_df, filtered <= 0.23) %>% slice_max(filtered)
```

```{r}
file <- "//wsl.localhost/Ubuntu/home/thomas/research/neoZ/2_heterozygosity/neosex.female_het.50kb.csv"
df <- read_csv(file, show_col_types = FALSE) %>%
  select(mid = mid_Mb, divergence = Female)

covered_wins <- median_coverage %>%
  filter(median > 0.84) %>% pull(mid)

windowed_divergence <- df %>%
  filter(mid %in% covered_wins)

res <- bcp(y = windowed_divergence$divergence, burnin = 500, mcmc = 5000)

win_div <- windowed_divergence %>%
    cbind(res$posterior.mean, res$posterior.prob) %>%
    rename(mean = X1, post = `res$posterior.prob`) %>%
    ggplot() +
    geom_point(aes(mid / 1e6, divergence), size = 1.2) +
    geom_line(aes(mid / 1e6, mean, col='red')) +
    labs(x = "Position (Mb)", y = "Divergence", title = "Neo-sex chromosome") +
    scale_x_continuous(breaks = seq(0, 12, 2)) +
    ylim(c(0, 0.035)) +
    annotate("segment", x = 0, xend = 10.45, y = 0.0025, yend = 0.0025, colour = "black", linewidth = 1) +
    annotate("text", x = 5.5, y = 0.004, label = "Plateau 1", size = 4) +
    annotate("segment", x = 10.55, xend = 12, y = 0.0025, yend = 0.0025, colour = "black", linewidth = 1) +
    annotate("text", x = 11.25, y = 0.004, label = "Plateau 2", size = 4) +
    annotate("segment", x = 12.0925, xend = 13, y = 0.0025, yend = 0.0025, colour = "black", linewidth = 1) +
    annotate("text", x = 12.6, y = 0.0012, label = "Plateau 3", size = 4) +
    theme_light() + 
    theme(legend.position = "None", plot.title = element_text(hjust = 0.5), 
          axis.title = element_text(size = 14))
    

win_div
```


(Supp Figure X)

```{r}
file <- "//wsl.localhost/Ubuntu/home/thomas/research/neoZ/2_heterozygosity/female.heterozigosity.50kb.csv"
df <- read_csv(file, show_col_types = FALSE) %>%
  rename(mid = mid_Mb)

df %>%
  left_join(meta, by = join_by(id)) %>%
  filter(mid %in% covered_wins, class == "Female") %>%
  ggplot(aes(col = id)) +
  geom_line(aes(mid / 1e6, heterozygosity), linewidth = 0.65, alpha = 0.6) +
  scale_x_continuous(breaks = seq(0, 12, 2)) +
  labs(x = "Position (Mb)", y = "Divergence", title = "Neo-sex chromosome", col = "Individual") +
  scale_color_manual(values = MetBrewer::met.brewer("Homer1", 7)) +
  geom_line(data = tibble(mean = c(res$posterior.mean), mid = windowed_divergence$mid/1e6), aes(mid, mean), col = "red", linewidth = 0.7) +
  theme_light() + 
  theme(plot.title = element_text(hjust = 0.5))
```

## Folded variant spectrum (Figure 5)

```{r}
tree<-read.tree(text="(((((1:140, 2:151):112, 3:259):259, 4:527):25, 5:523):32, (6:563, 7:576):9);")
tree<-read.tree(text="(((((1:140, 2:151):112, 3:259):259, 4:527):250, 5:523):320, (6:563, 7:576):1);")
colours <- c(rainbow(12))
colours[5] <- "darkolivegreen1"
colours[6] <- "violet"
tree$tip.label <- c("PT_MI_1680 (1)", "ES_MI_1684 (7)", "PT_MI_86 (5)", "ES_MI_1683 (4)", "ES_MI_1682 (2)", "PT_MI_61 (3)", "ES_MI_1686 (6)")

tree<-read.tree(text="(((((1:140, 2:151):112, 3:259):259, 4:527):25, 5:523):32, (6:563, 7:576):9);")
tcols <- c(rainbow(13))
tcols[5] <- "darkolivegreen1"
tcols[10] <- "blue4"
tree$tip.label <- c("PT_MI_1680 (1)", "ES_MI_1684 (7)", "PT_MI_86 (5)", "ES_MI_1683 (4)", "ES_MI_1682 (2)", "PT_MI_61 (3)", "ES_MI_1686 (6)")
ptree <- ggtree(tree, aes(color = factor(node)), size = 2) + 
  geom_tiplab(hjust = -0.1) + 
  scale_color_manual(values=tcols[c(5, 6, 7, 8, 9, 11, 12, 3, 1, 2, 3, 4, 1)]) + 
  theme_tree2() + 
  theme(legend.position = "none") + 
  xlim(c(0, 750)) + 
  xlab("Derived mutations") +
  theme(axis.title = element_text(size = 18), 
        axis.line = element_line(colour = 'black', size = 0.75),
        axis.text = element_text(size = 13))


file <- "//wsl.localhost/Ubuntu/home/thomas/research/neoZ/020_final/neoZ_folded_SNPs_type.csv"
neoz <- read_csv(file, show_col_types = FALSE) %>%
  rename("Neo-Z" = freq, "Expected" = exp_freq)

file <- "//wsl.localhost/Ubuntu/home/thomas/research/neoZ/020_final/neoW_folded_SNPs_type.csv"
neow <- read_csv(file, show_col_types = FALSE) %>%
  select(pol, iton, "Neo-W" = freq)

df <- left_join(neoz, neow, by = join_by(pol, iton)) %>%
  replace(is.na(.), 0)

colours <- c(c(tcols[6], tcols[12], tcols[7], tcols[8], tcols[11], tcols[9], tcols[5]), rep("darkgrey", 7), tcols[1], rep("darkgrey", 7), tcols[4], rep("darkgrey", 22), tcols[2], rep("darkgrey", 3), tcols[3], rep("darkgrey", 30))

var_types <- df %>% arrange(iton) %>%
  mutate(id = row_number()) %>%
  pivot_longer(cols = c("Neo-Z", "Neo-W", "Expected")) %>%
  ggplot(aes(x = forcats::fct_reorder(pol, id), y = value, fill = name)) +
  geom_col(position = "dodge", col = "gray") +
  scale_fill_manual(values = c("orangered", "#FFFFC1", "#FFD1D1")) +
  labs(x = "Variant type", y = "Frequency", fill = "") +
  theme_light() +
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1, color=colours, size = 12),
        axis.text.y = element_text(size = 13),
        axis.title = element_text(size = 18), 
        axis.title.x = element_text(vjust = 0.1))

folded <-  var_types / ptree + plot_annotation(tag_levels = 'A')
folded
```


## Degeneration (Figure 6)

### Gene expression


```{r}
#file <- "//wsl.localhost/Ubuntu/home/thomas/research/neoZ/017_RNAseq/neoW.pseudoref.gene.pos.read_count.txt"
file <- "//wsl.localhost/Ubuntu/home/thomas/research/neoZ/017_RNAseq/neoW.final.pseudoref.gene.pos.read_count.txt"
neow <- read_delim(file, delim = "\t", col_names = c("id", "neow"), show_col_types = FALSE)

file <- "//wsl.localhost/Ubuntu/home/thomas/research/neoZ/017_RNAseq/neoZ.gene.pos.read_count.txt"
neoz <- read_delim(file, delim = "\t", col_names = c("id", "neoz"), show_col_types = FALSE) 
nznw_expression <- left_join(neow, neoz, by = join_by(id))

file <- "//wsl.localhost/Ubuntu/home/thomas/research/neoZ/1_clean_vcfs/gene_annotation/neo_sex.genes.plateaus.bed"
plateaus <- read_delim(file, delim = "\t", col_names = c("plateau", "gene"), show_col_types = FALSE) %>%
  mutate(gene = str_remove(gene, "ID="), 
         plateau = str_replace(plateau, "plat_", "Plateau "))

expression <- nznw_expression %>%
  left_join(plateaus, by = c("id" = "gene")) %>%
  drop_na() %>%
  filter(neow + neoz > 3) %>%
  #filter(nw > 5, nz > 5) %>%
  mutate(change = neoz / neow, 
         change = ifelse(change == Inf, 1000, ifelse(change == 0, 0.001, change)),
         change = log(change, 2)) %>%
  ggplot(aes(change)) +
  geom_histogram(bins = 25) +
  geom_vline(xintercept = 0, col = "red", linetype = "dashed") +
  labs(x = expression(log[2](Neo-Z / Neo-W)), y = "Frequency") +
  scale_x_continuous(labels = c(expression(-infinity), -5, 0, 5, expression(infinity))) +
  facet_grid(~plateau) +
  theme_light()


conf <- 0.95

lof <- tibble(plat = rep(c(1, 2, 3), 3), genes = rep(c(222, 28, 25), 3),
       name = c("Neo-W", "Neo-W", "Neo-W", "Neo-Z", "Neo-Z", "Neo-Z", "North-Africa", "North-Africa", "North-Africa"),
       value = c(59, 10, 1, 3, 2, 1, 14, 4, 2)) %>%
  rowwise() %>%
  mutate(plat = paste("Plateau", plat), 
         prop = value / genes, 
         low = binom.test(value, genes, conf.level = conf)$conf.int[1], 
         high = binom.test(value, genes, conf.level = conf)$conf.int[2]) %>%
  ggplot(aes(plat, prop, fill = name)) +
  geom_col(position = "dodge", col = "black") +
  geom_errorbar(aes(ymin = low, ymax = high), position = position_dodge(0.9), width = 0.15, linewidth = 0.58) +
  lims(y = c(0, 0.6)) +
  scale_fill_manual(values = c("#FFFFC1", "#FFD1D1", "#EA927A")) +
  labs(x = "", y = "% of genes with fixed loss-of-function mutations", fill = "") +
  theme_light()

degeneration <- (expression | lof) + plot_annotation(tag_levels = 'A')
degeneration
```

```{r}
ggsave("pngs/neoW.degeneration.final1.png", degeneration, width = 10, height = 5, dpi = 500)
```

```{r}
ddf <- nznw_expression %>%
  left_join(plateaus, by = c("id" = "gene")) %>%
  drop_na() %>%
  filter(neow + neoz > 3) %>%
  mutate(change = neoz / neow, 
         change = ifelse(change == Inf, 1000, ifelse(change == 0, 0.001, change)),
         lchange = log(change, 2))

median(ddf$change)
```